pdns/recursordist/rec-system-resolve.cc

```diff
@@ -0,0 +1,292 @@
+/*
+ * This file is part of PowerDNS or dnsdist.
+ * Copyright -- PowerDNS.COM B.V. and its contributors
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of version 2 of the GNU General Public License as
+ * published by the Free Software Foundation.
+ *
+ * In addition, for the avoidance of any doubt, permission is granted to
+ * link this program with OpenSSL and to (re)distribute the binaries
+ * produced as the result of such linking.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netdb.h>
+#include <arpa/nameser.h>
+#include <resolv.h>
+
+#include "dnsparser.hh"
+#include "dnsrecords.hh"
+#include "rec-system-resolve.hh"
+#include "logging.hh"
+#include "noinitvector.hh"
+#include "threadname.hh"
+
+namespace
+{
+ComboAddress resolve(const std::string& name)
+{
+  struct addrinfo hints = {};
+  hints.ai_flags = AI_ADDRCONFIG;
+  hints.ai_family = 0;
+
+  struct addrinfo* res = nullptr;
+  auto ret = getaddrinfo(name.c_str(), nullptr, &hints, &res);
+  // We pick the first address returned for now.
+  // XXX This might trigger unwanted "changes detected "if the sort order varies
+  if (ret == 0) {
+    auto address = ComboAddress{res->ai_addr, res->ai_addrlen};
```

Comment:
It looks like we will leak `res` if `ComboAddress`'s ctor throws (unlikely, granted).